### PR TITLE
Mitigate overhead of stack emptyness checks in Ingredient

### PR DESCRIPTION
**Pull request by embeddedt**
We are seeing the overhead of the patched-in logic to this method appearing quite often in FC profiles. Until this system can be overhauled properly, this PR aims to mitigate the overhead by caching the result of the emptiness check once, and reusing it on future invocations to `isEmpty`. I also chose to get rid of the stream since it probably isn't helping matters if there are very few stacks per ingredient.

I would appreciate feedback from @marchermans in particular on whether this solution is correct, or whether the underlying stacks could change.

### CODE REVIEW

**Comment by pupnewfster** [comment]
Why the extra empty local variable? From what I can tell it isn’t used and could just use the field?

**Comment by Technici4n** [comment]
This is fine and will maintain current behavior. We'll really have to clean this up in 1.20.5.

**Review by marchermans — DISMISSED** [review]
Sadly this is not really a great idea.
We had this system in the past (check 20.1) and we ran into issues with Tags not properly working, or dynamic ingredients not properly handling it. We introduced a cache busting mechanism, but all of that is immensely painful to maintain. So in 20.2 we ripped that all out in favor of that check.

I would like to see some numbers here, but instead of caching it and introducing that nightmare of code, I would first get rid of the streams call.....

**Comment by embeddedt** [comment]
I can't find an `isEmpty` patch in 1.20.1, could you point me to where that's being done? Also I'm a bit confused how this is a "nightmare" of code... it's just an unrolling of the stream into a regular loop + caching the result of that in a field.

**Comment by TelepathicGrunt** [comment]
@marchermans @embeddedt The isEmpty thing came up before.
https://discord.com/channels/313125603924639766/852298000042164244/1177968547939487826

Tldr: isEmpty patch was added in 1.20.2 porting. The patch to ShapelessRecipe to not call isEmpty is what was pre-existing in past. Modders calling the ingredient's isEmpty method too early will kill the tag permanently for the lifetime of the game whereas vanilla behavior can have the tag recover later without issue. Making this early isEmpty call a nightmare to debug for modders as there's zero identifying info of what the problem is when it happens...

**Comment by pupnewfster** [comment]
Orion is talking about this set of is the items and stuff for the ingredient valid: https://github.com/neoforged/NeoForge/blob/1.20.1/patches/minecraft/net/minecraft/world/item/crafting/Ingredient.java.patch#L48-L64

**Comment by Technici4n** [comment]
@marchermans the stacks never change once the ingredient is dissolve. I don't see why the isEmpty check would have to change.

File: patches/net/minecraft/world/item/crafting/Ingredient.java.patch [hunk 1]
```diff
@@ -78,11 +79,26 @@
 +     */
 +    public boolean synchronizeWithContents() {
 +        return true;
++    }
++
++    private boolean areAllStacksEmpty() {
++        Boolean empty = this.areAllStacksEmpty;
```
Comment (pupnewfster) [inline patches/net/minecraft/world/item/crafting/Ingredient.java.patch:85]: I believe rather than having this local variable, we can just check `if (this.areAllStacksEmpty == null)` and then do `this.areAllStacksEmpty = allEmpty;` and `return this.areAllStacksEmpty`?

**Review by pupnewfster — CHANGES_REQUESTED** [review]

File: patches/net/minecraft/world/item/crafting/Ingredient.java.patch (hunk 1 above)
Comment (embeddedt) [reply patches/net/minecraft/world/item/crafting/Ingredient.java.patch:85]: I used this coding pattern to avoid reading the non-final field multiple times. May not make a difference to the JIT's ability to optimize, but doesn't hurt IMO.

**Review by pupnewfster — APPROVED** [review]

File: patches/net/minecraft/world/item/crafting/Ingredient.java.patch (hunk 1 above)
Comment (Technici4n) [reply patches/net/minecraft/world/item/crafting/Ingredient.java.patch:85]: That pattern is generally only used for volatile variables, or in case of concurrent mutations.

**Comment by marchermans** [comment]
Ignore my comment
I misread the patch on my phone
